### DISCUSSION TITLE
GitHub Packages | Support for Dart/Flutter packages

### DISCUSSION THREAD

**Discussion by AesSedai101** [community/community#8536] [Packages]
Right now, the only option for Dart/Flutter packages is to publish them publicly to pub.dev or to set up your own server.

I would like to deploy my internal Dart packages to GitHub packages privately so that they can be managed alongside my other packages.

**Comment by lucasilverentand** [comment]
Hi, I very much would like to see this feature too. I'm currently looking into setting up an unpub server but would rather have the packages living in our private repositories themselves through GitHub Packages. How hard would it actually be to implement this?

**Comment by tmorone-rezi** [comment]
+1. Flutter is only gaining in popularity. We are currently evaluating jFrog Artifactory because they support pub packages. Would love to see this in Github package repository.

**Comment by m-skolnick** [comment]
For now, a good alternative is Cloudsmith. They have full dart/flutter support, and now dart supports hosted urls in addition to pub.dev as well.

**Comment by iamchathu** [comment]
Currently JetBrains Spaces, OnePub.dev supports dart/Flutter package registry and they have a free tier.  Also all listed in here https://dart.dev/tools/pub/custom-package-repositories#dart-package-repositories-as-a-service. I'd like to see GitHub Package registry to have this feature.

**Comment by carmonac** [comment]
any news about this?

**Comment by rxseb** [comment]
I would love that as well.

**Comment by Albertbol** [comment]
up!

**Comment by sigurdm** [comment]
The pub client protocol is documented here: https://github.com/dart-lang/pub/blob/master/doc/repository-spec-v2.md .

If there is something we can do to support this from the Dart Pub team, please get in touch!

**Comment by alexdecoder** [comment]
I would love to see this as well. Flutter continues to build in popularity, and it will become increasingly important for this to be answered, My organization is evaluating alternatives as a result of this missing functionality.

**Comment by westito** [comment]
You can simply use plain git repos for package source.

**Reply by thecodemonk** [reply]
Except it's a pain in the butt to do that. If you don't specify some reference it till never retrieve another update for that repo.

**Reply by westito** [reply]
I didn't notice. Usually pub get updates changed repos even no reference given. But, as you changing versions with pub.dev packages when upgrading, you can change git refs also. It's the same method.

**Comment by klaas-videc** [comment]
You would not be able to use constraints on your "git-only-packages" though, would you?

**Reply by westito** [reply]
Git package have the same package format as the pub.dev packages. The only difference is the code location (that can be a local folder too). You can specify a branch/commithash/tag for versioning. I use separate git repos, submodules and local folders too as custom Dart package. Works like a charm. 
https://dart.dev/tools/pub/dependencies#git-packages

**Reply by klaas-videc** [reply]
You are right. What i meant was, resolving conflicting constraints for my git-repo-package would not implicitly happen since the pub-system does not know a list of available versions for my git-repo-package.

**Reply by westito** [reply]
Indeed. Semantic versioning and resolution won't work, it's more like a dependency override. I use it in private projects where each repo's actual state is always consitent, so semantic versioning and dependency resolution is irrelevant. However, you can override the git-sourced sub-packages in root project if necessary just like other packages. For ex. if you have a custom framework as private git package and you want to update a single package inside the framework just for a certain project.

**Reply by sigurdm** [reply]
Actually we are working on this (don't know when exactly (or even if) this will land...)
https://github.com/dart-lang/pub/pull/4427

**Comment by tsinis** [comment]
Hey GitHub Packages Team!

Is there any way the Dart community can help unblock this? The core registry is closed-source (?), but if there are any integrations, Actions, or tooling surfaces where we can submit PRs to help build this out, we just need the guidance, and we will do the work.

FWIW, GitLab is currently working on this and has an active WIP Merge Request open for a native Dart registry ([gitlab-org/gitlab!217611](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/217611)). We'd love to see GitHub get there first (or at least match it!).

What can we do to help make this happen? 🤔
